src/net/db.rs

```diff
@@ -1073,6 +1283,136 @@ impl PeerDB {
+        conn: &Connection,
+        network_id: u32,
+        addrbytes: &PeerAddress,
+        port: u16,
+    ) -> Result<Option<u32>, db_error> {
+        // the peer can be at any of these slots, if it is present at all.
+        let slots = PeerDB::peer_slots(conn, network_id, addrbytes, port)?;
+        for slot in slots.into_iter() {
+            let used_slot = PeerDB::has_specific_peer_at(conn, network_id, addrbytes, port, slot)?;
+            if used_slot {
+                return Ok(Some(slot));
+            }
+        }
+        Ok(None)
+    }
+
+    /// Get the list of stacker DB contract IDs for a given set of slots.
+    /// The list will contain distinct contract IDs.
+    fn get_stacker_dbs_by_slot(
+        conn: &Connection,
+        used_slot: u32,
+    ) -> Result<Vec<ContractId>, db_error> {
+        let mut db_set = HashSet::new();
+        let qry = "SELECT smart_contract_id FROM stackerdb_peers WHERE peer_slot = ?1";
+        let dbs = query_rows(conn, qry, &[&used_slot])?;
+        for cid in dbs.into_iter() {
+            db_set.insert(cid);
+        }
+
+        Ok(db_set.into_iter().collect())
+    }
+
+    /// Get the slots for all peers that replicate a particular stacker DB
+    fn get_stacker_db_slots(
+        conn: &Connection,
+        smart_contract: &ContractId,
+    ) -> Result<Vec<u32>, db_error> {
+        let qry = "SELECT peer_slot FROM stackerdb_peers WHERE smart_contract_id = ?1";
+        let args: &[&dyn ToSql] = &[&smart_contract.to_string()];
+        query_rows(conn, qry, args)
+    }
+
+    /// Get a peer's advertized stacker DBs
+    fn static_get_peer_stacker_dbs(
+        conn: &Connection,
+        neighbor: &Neighbor,
+    ) -> Result<Vec<ContractId>, db_error> {
+        let used_slot_opt = PeerDB::find_peer_slot(
+            conn,
+            neighbor.addr.network_id,
+            &neighbor.addr.addrbytes,
+            neighbor.addr.port,
+        )?;
+        if let Some(used_slot) = used_slot_opt {
+            Self::get_stacker_dbs_by_slot(conn, used_slot)
+        } else {
+            Ok(vec![])
+        }
+    }
+
+    /// Get a peer's advertized stacker DBs by their IDs.
+    pub fn get_peer_stacker_dbs(&self, neighbor: &Neighbor) -> Result<Vec<ContractId>, db_error> {
+        PeerDB::static_get_peer_stacker_dbs(&self.conn, neighbor)
+    }
+
+    /// Update an existing peer's stacker DB IDs.
+    /// Calculates the delta between what's in the DB now, and what's in `dbs`, and deletes the
+    /// records absent from `dbs` and adds records not present in the DB.
+    /// Does nothing if the peer is not present.
+    pub fn update_peer_stacker_dbs<'a>(
+        tx: &Transaction<'a>,
+        neighbor: &Neighbor,
+        dbs: &[ContractId],
+    ) -> Result<(), db_error> {
+        let cur_slot_opt = PeerDB::find_peer_slot(
+            tx,
+            neighbor.addr.network_id,
+            &neighbor.addr.addrbytes,
+            neighbor.addr.port,
+        )?;
```

Comment:
If there's no slot, shouldn't this function do nothing?

```suggestion
        let neighbor_slot = match PeerDB::find_peer_slot(
            tx,
            neighbor.addr.network_id,
            &neighbor.addr.addrbytes,
            neighbor.addr.port,
        )? {
            Some(x) => x,
            None => return Ok(()),
        };
```

Reply:
Sure thing